04_basic_struct/module_basic_struct/basic_struct.c

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/module.h>
+#include <linux/slab.h>
+
+struct message_list {
+	struct list_head entry;
+	char *buffer;
+};
+
+static struct list_head my_list;
+
+static ssize_t list_show(struct kobject *kobj, struct kobj_attribute *attr,
+	char *buf)
+{
+	struct message_list *pos;
+	char *p = buf, *prehead = "Constent of list";
+	ssize_t ret, size, size_left = PAGE_SIZE; // sysfs allocated for us whole page
+	int i = 0;
+
+	list_for_each_entry(pos, &my_list, entry) {
+		// size of one list element with head message
+		size = strlen(pos->buffer) + strlen(prehead) + 4;
+		// copy into buffer formatted string
+		ret = scnprintf(p, size_left, "%s %d: %s", prehead, i++, pos->buffer);
+		if (ret != size)
+			return -ENOMEM;
+
+		size_left -= size;
+		p += size;
+	}
+
+	return PAGE_SIZE - size_left;
+}
+
+static ssize_t list_store(struct kobject *kobj, struct kobj_attribute *attr,
+	const char *buf, size_t count)
+{
+	char *new_msg;
+	struct message_list *new_entry;
+
+	new_msg = kzalloc(count, GFP_KERNEL);
+	if (!new_msg)
+		return -ENOMEM;
+
+	memcpy(new_msg, buf, count);
+
+	new_entry = kzalloc(sizeof(*new_entry), GFP_KERNEL);
+	if (!new_entry)
+		return -ENOMEM;
```

Comment:
Before return memory freeing for new_msg is required.

Reply:
Fixed in 2e684c7